### PR TITLE
fix isa tfunc for type kinds

### PR DESCRIPTION
**Pull request by vtjnash**
previously, isa_tfunc(DataType, Type{...}) might return Const(false)

### CODE REVIEW

**Comment by Keno** [comment]
Add some version of the original problematic code to the test?

**Comment by vtjnash** [comment]
It's in there; I just reduced it to the precise MWE and unit-tested that.

**Comment by Keno** [comment]
You're whitebox testing tfuncs though. I'd much prefer a test along the lines of the original code.

File: base/inference.jl [hunk 1]
```diff
@@ -378,19 +378,21 @@ isType(t::ANY) = isa(t, DataType) && (t::DataType).name === _Type_name
...
-const n_ifunc = reinterpret(Int32,arraylen)+1
-const t_ifunc = Array{Tuple{Int,Int,Any},1}(n_ifunc)
-const t_ifunc_cost = Array{Int,1}(n_ifunc)
-const t_ffunc_key = Array{Function,1}(0)
-const t_ffunc_val = Array{Tuple{Int,Int,Any},1}(0)
-const t_ffunc_cost = Array{Int,1}(0)
+const n_ifunc = reinterpret(Int32, arraylen) + 1
+const t_ifunc = Array{Tuple{Int, Int, Any}, 1}(n_ifunc)
+const t_ifunc_cost = Array{Int, 1}(n_ifunc)
+const t_ffunc_key = Array{Any, 1}(0)
```
Comment (JeffBezanson) [inline base/inference.jl]: Are we putting an non-Functions in here?

Comment (vtjnash) [inline base/inference.jl]: Nah – actually we only put `Builtin` subtypes in here. This should just be more likely to trigger our anti-specialization heuristics and avoid trying to put convert and typeassert into the code path.

**Comment by vtjnash** [comment]
> You're whitebox testing tfuncs though. I'd much prefer a test along the lines of the original code.

I think you mean unit-testing, and that you would like to also see an integration test? Adding a test for the original failure would also be a whitebox test. However, I'm not sure that the integration test is necessary in this case. We already have a couple tests in this file to show that integration works, so I think we were just missing some of the cases for unit testing.

**Comment by JeffBezanson** [comment]
@vtjnash I suspect this might not backport cleanly, and you might need to rebase it for 0.6 (and almost certainly for 0.5) manually.

**Comment by ararslan** [comment]
@vtjnash I'm not sure how to backport this properly to 0.6. Could you submit this change as a PR to the release-0.6 branch?

**Pull request merged**